src/engine/MaxConstraint.cpp

```diff
@@ -85,35 +85,39 @@ void MaxConstraint::restoreState( const PiecewiseLinearConstraint *state )
 
 void MaxConstraint::registerAsWatcher( ITableau *tableau )
 {
-    tableau->registerToWatchVariable( this, _f );
     for ( unsigned element : _elements )
-    {
-        if ( element == _f )
-            continue;
         tableau->registerToWatchVariable( this, element );
-    }
+
+    if ( !_elements.exists( _f ) )
+        tableau->registerToWatchVariable( this, _f );
 }
 
 void MaxConstraint::unregisterAsWatcher( ITableau *tableau )
 {
-    tableau->unregisterToWatchVariable( this, _f );
+    // Guy: I think there's a bug here, not related to the recent changes:
+    // the list of _elements can change, and so some of them might not
+    // get propertly unregistered.
     for ( unsigned element : _elements )
-    {
-        if ( element == _f )
-            continue;
         tableau->unregisterToWatchVariable( this, element );
-    }
+
+    if ( !_elements.exists( _f ) )
+        tableau->unregisterToWatchVariable( this, _f );
 }
 
 void MaxConstraint::notifyVariableValue( unsigned variable, double value )
 {
+    /*
+      Guy: I'm not sure I understand this condition. If _f is an
+      element, what is the _maxIndex? Is it always _f? Is it the
+      highest element if that's not _f?
+    */
```

Comment:
I assume:
If _f is not an element, _maxIndex is the highest element of _elements.
If _f is an element, maxIndex is the highest element of _elements including _f.

Namely, _maxIndex is always the highest element of elements.

Reply:
Okay.